### PR TITLE
[ARM] Cross Tenant Support for group deployment

### LINKED ISSUE #11151: Support Cross-Tenant for az group deployment (arm template deployments) by taking auxilliary-tenant-id in input
**Resource Provider** 
Microsoft.Compute, ARM

**Description of Feature or Work Requested**
Today, with SIG we support cross tenant vm deployment, i.e. the share image gallery (sig) is in different tenant than the vm being deployed to. This is only supported in az vm create. The way it works is cli takes a token from sig's tenant and also from vm's tenant (target) and passes both of them to ARM.
But this confuses the customers today, as they expect it to work from az group deployment (template deployments) also, but it doesnt work. It might be difficult to parse template, so that's why I request cli team supports an auxilliary tenant id in input to the az group deployment, and passes along the token to arm along with template deployment create request.

**Minimum API Version Required**
Not related

**Swagger Link**
N/A

**Target Date**
It is causing customer pain, and confusion. It will also unblock a major scneario in azure, and the work might turn out to be small. If it is a low hanging fruit, we will really appreciate your help to ease the customer pain.

### PR DESCRIPTION
**Pull request by Juliehzl**
Fix #11151
---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

**Comment by Juliehzl** [comment]
Based on a SIG created in another tenant, live run test can pass locally.

File: src/azure-cli/azure/cli/command_modules/resource/custom.py [hunk 1]
```diff
@@ -925,13 +926,16 @@ def delete_deployment_at_subscription_scope(cmd, deployment_name):
...
     if handle_extended_json_format:
         return _deploy_arm_template_unmodified(cmd.cli_ctx, resource_group_name, template_file, template_uri,
-                                               deployment_name, parameters, mode, rollback_on_error, no_wait=no_wait)
+                                               deployment_name, parameters, mode, rollback_on_error, no_wait=no_wait,
+                                               aux_subscriptions=aux_subscriptions)
 
     return _deploy_arm_template_core(cmd.cli_ctx, resource_group_name, template_file, template_uri,
-                                     deployment_name, parameters, mode, rollback_on_error, no_wait=no_wait)
+                                     deployment_name, parameters, mode, rollback_on_error, no_wait=no_wait,
+                                     aux_subscriptions=aux_subscriptions)
```
Comment (yungezz) [inline src/azure-cli/azure/cli/command_modules/resource/custom.py]: So arm service will find tenant of those subs themselves?

File: src/azure-cli/azure/cli/command_modules/resource/custom.py [hunk 2]
```diff
@@ -925,13 +926,16 @@ def delete_deployment_at_subscription_scope(cmd, deployment_name):
 
 def deploy_arm_template(cmd, resource_group_name,
                         template_file=None, template_uri=None, deployment_name=None,
-                        parameters=None, mode=None, rollback_on_error=None, no_wait=False, handle_extended_json_format=False):
+                        parameters=None, mode=None, rollback_on_error=None, no_wait=False,
+                        handle_extended_json_format=False, aux_subscriptions=None):
     if handle_extended_json_format:
         return _deploy_arm_template_unmodified(cmd.cli_ctx, resource_group_name, template_file, template_uri,
-                                               deployment_name, parameters, mode, rollback_on_error, no_wait=no_wait)
+                                               deployment_name, parameters, mode, rollback_on_error, no_wait=no_wait,
```
Comment (yungezz) [inline src/azure-cli/azure/cli/command_modules/resource/custom.py]: avoid positional params

File: src/azure-cli/azure/cli/command_modules/resource/custom.py (hunk 1 above)
Comment (Juliehzl) [reply src/azure-cli/azure/cli/command_modules/resource/custom.py]: CLI will find all information including tenant id information according the subscription through [get_subscription()](https://github.com/Azure/azure-cli/blob/dev/src/azure-cli-core/azure/cli/core/_profile.py#L484) in [here](https://github.com/Azure/azure-cli/blob/dev/src/azure-cli-core/azure/cli/core/_profile.py#L535) and get the access token for the resource.

File: src/azure-cli/azure/cli/command_modules/resource/custom.py (hunk 2 above)
Comment (Juliehzl) [reply src/azure-cli/azure/cli/command_modules/resource/custom.py]: Got it

File: src/azure-cli/azure/cli/command_modules/resource/_params.py [hunk 3]
```diff
@@ -172,16 +172,23 @@ def load_arguments(self, _):
...
+        c.argument('rollback_on_error', nargs='?', action=RollbackAction,
+                   help='The name of a deployment to roll back to on error, or use as a flag to roll back to the last '
+                        'successful deployment.')
 
     with self.argument_context('group deployment create') as c:
         c.argument('deployment_name', options_list=['--name', '-n'], required=False,
                    help='The deployment name. Default to template file base name')
         c.argument('handle_extended_json_format', arg_type=extended_json_format_type)
+        c.argument('aux_subscriptions', nargs='*', options_list=['--aux-subs'],
+                   help='Auxiliary subscriptions which wil be used during deployment in cross tenants.')
```
Comment (qwordy) [inline src/azure-cli/azure/cli/command_modules/resource/_params.py]: wil -> will

File: src/azure-cli/azure/cli/command_modules/resource/tests/latest/crosstenant_vm_deploy.json [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+                    "networkInterfaces": [
+                        {
+                            "id": "[parameters('NIC_id')]"
+                        }
+                    ]
+                },
+                "storageProfile": {
+                    "osDisk": {
+                        "createOption": "fromImage",
+                        "name": None,
```
Comment (qwordy) [inline src/azure-cli/azure/cli/command_modules/resource/tests/latest/crosstenant_vm_deploy.json]: None is highlighted RED. Is it an error which Github can't stand?

File: src/azure-cli/azure/cli/command_modules/resource/tests/latest/test_resource.py [hunk 5]
```diff
@@ -438,6 +438,47 @@ def test_group_deployment(self, resource_group):
             self.check('[0].resourceGroup', '{rg}')
         ])
 
+    @ResourceGroupPreparer(name_prefix='cli_test_deployment')
+    def test_group_deployment(self, resource_group):
```
Comment (MyronFanQiu) [inline src/azure-cli/azure/cli/command_modules/resource/tests/latest/test_resource.py]: Both two tests are live_only. Right?

File: src/azure-cli/azure/cli/command_modules/resource/_params.py [hunk 6]
```diff
@@ -172,16 +172,23 @@ def load_arguments(self, _):
     with self.argument_context('group deployment') as c:
         c.argument('resource_group_name', arg_type=resource_group_name_type, completer=get_resource_group_completion_list)
         c.argument('deployment_name', options_list=['--name', '-n'], required=True, help='The deployment name.')
-        c.argument('template_file', completer=FilesCompleter(), type=file_type, help="a template file path in the file system")
+        c.argument('template_file', completer=FilesCompleter(), type=file_type,
+                   help="a template file path in the file system")
```
Comment (qwordy) [inline src/azure-cli/azure/cli/command_modules/resource/_params.py]: I prefer removing "a" or adding "the"

File: src/azure-cli/azure/cli/command_modules/resource/_params.py (hunk 3 above)
Comment (qwordy) [reply src/azure-cli/azure/cli/command_modules/resource/_params.py]: in cross tenants -> across tenants

File: src/azure-cli/azure/cli/command_modules/resource/_params.py [hunk 7]
```diff
@@ -172,16 +172,23 @@ def load_arguments(self, _):
...
         c.argument('template_uri', help='a uri to a remote template file')
```
Comment (qwordy) [inline src/azure-cli/azure/cli/command_modules/resource/_params.py]: a -> an, uri -> URI

File: src/azure-cli/azure/cli/command_modules/resource/_params.py [hunk 8]
```diff
@@ -172,16 +172,23 @@ def load_arguments(self, _):
...
-        c.argument('mode', arg_type=get_enum_type(DeploymentMode, default='incremental'), help='Incremental (only add resources to resource group) or Complete (remove extra resources from resource group)')
+        c.argument('mode', arg_type=get_enum_type(DeploymentMode, default='incremental'),
```
Comment (qwordy) [inline src/azure-cli/azure/cli/command_modules/resource/_params.py]: incremental -> Incremental

Comment (Juliehzl) [reply src/azure-cli/azure/cli/command_modules/resource/_params.py]: It is not case sensitive

File: src/azure-cli/azure/cli/command_modules/resource/tests/latest/crosstenant_vm_deploy.json (hunk 4 above)
Comment (Juliehzl) [reply src/azure-cli/azure/cli/command_modules/resource/tests/latest/crosstenant_vm_deploy.json]: For JSON format, None value is not supported but it doesn't matter in ARM deployment. But I can give it valid value for better description and validity.

File: src/azure-cli/azure/cli/command_modules/resource/tests/latest/test_resource.py (hunk 5 above)
Comment (Juliehzl) [reply src/azure-cli/azure/cli/command_modules/resource/tests/latest/test_resource.py]: Thanks for reminder! It is duplicated and should be deleted!

File: src/azure-cli/azure/cli/command_modules/resource/_params.py (hunk 6 above)
Comment (qwordy) [reply src/azure-cli/azure/cli/command_modules/resource/_params.py]: You can leave a reply before resolving it. Even if you think my comment is sheer nonsense, please respond. 🙂

**Pull request merged**